### ISSUE TITLE
Typing : * $ @ etc in emmet abbreviation stops showing suggestions

### ISSUE THREAD

**Issue by jens1o** [Microsoft/vscode#29758]
- VSCode Version: Code - Insiders 1.14.0-insider (63def6eb989d62f23d3711ace0f795ce9f23551d, 2017-06-28T05:16:57.588Z)
- OS Version: Windows_NT ia32 10.0.15063
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
Bookmarks|ale|0.15.2
path-intellisense|chr|1.4.2
gitignore|cod|0.5.0
gitlens|eam|4.2.0
tslint|eg2|0.15.0
LogFileHighlighter|emi|1.1.1
vscode-reveal|evi|0.0.9
php-debug|fel|1.10.1
php-intellisense|fel|1.4.1
auto-close-tag|for|0.4.2
gc-excelviewer|Gra|1.1.15
composer|ika|0.5.0
smarty|imp|0.2.0
json-to-ts|Mar|1.4.3
vscode-apache|mrm|1.1.1
cpptools|ms-|0.12.0
github-issues-prs|ms-|0.2.0
php-docblocker|nei|1.2.0
vscode-code-outline|pat|0.0.6
vscode-versionlens|pfl|0.19.1
vscode-icons|rob|7.10.1
sharecode|Rol|0.4.1
code-spell-checker|str|1.2.0
lorem-ipsum|Tyr|1.0.0
vscode-todo-highlight|way|0.5.5
highlight-trailing-white-spaces|yba|0.0.2

(2 theme extensions excluded)

---

Steps to Reproduce:

1. Create a new `index.html` file
2. Write `html`, see there is a `html:5` suggestion
3. Type `:`, they are gone. 😢 

Reproduces without extensions: Yes

**Comment by ramya-rao-a** [comment]
@jrieken Looks like something changed in the filtering logic of suggestions. Your recent change around the `inComplete:true` maybe?

The emmet completion provider returns a completion list with `inComplete` set to `true`.
In stable, every letter typed would call the completion provider, but `:` would just filter existing results.

Stable:

![stablefilter](https://user-images.githubusercontent.com/16890566/27666776-13fbb958-5c2b-11e7-981b-b8822bea34ba.gif)

In the current Insiders, every letter typed still calls the completion provider, but `:` filters out everything

Current Insiders:

![currentfilter](https://user-images.githubusercontent.com/16890566/27666774-12504d4e-5c2b-11e7-8dd1-feee2e32a35e.gif)

**Comment by jens1o** [comment]
Yes, I was curious in the first moment why it stopped working.

**Comment by jrieken** [comment]
Oh, bummer... That's a regression from a change I made yesterday to fix #28400. I'll might revert that change...

**Comment by jens1o** [comment]
Sorry to bother you. :/

**Comment by jrieken** [comment]
Hm, reverting https://github.com/Microsoft/vscode/commit/5ad16183ab248948fdc457af0340321caf1ba987 (which is the fix for #28400) doesn't help. @ramya-rao-a It seems that `:` is registered as a trigger character and when we see that we invoke the provider in question and merge its results with the existing result. No, the problem is that that provider doesn't return anything and that the existing results don't match anymore... 

<img width="1010" alt="screen shot 2017-06-29 at 09 32 03" src="https://user-images.githubusercontent.com/1794099/27676827-c6cc618e-5caf-11e7-82c3-8e653a2dee35.png">

**Comment by jrieken** [comment]
> @jrieken Looks like something changed in the filtering logic of suggestions. Your recent change around the inComplete:true maybe?

So, we did make a change yesterday (I think too early for the version this was reported for) which does the following: When a suggestion result is incomplete *and* if the word as stopped we *don't* re-trigger suggestions but filter the suggestions we currently have. So, when typing `html:` the word ends and we stop asking for new suggestions (that might be a problem with expansions form `html:foo"bar` but I am unsure that exists). However, since the results already contain `html:5` it will be shown and can be selected. 

I wonder if this was caused by having another provider that registers with the `:` trigger character (I believe https://github.com/Microsoft/vscode/commit/7ba1dcff0c5f89c4582598eef4e05830d4dce306). So what happens, and I don't think much has changed there, is that hitting a trigger-character only asks that provider (which doesn't return a result). Its (empty) result is being merged with the existing completions (from the incomplete provider) which didn't get a chance to update and hence fail the filter condition. 

In short, there is something off. I am not yet convinced that a recent change caused this but things are complex. I acknowledge that there is a race between incomplete suggestions and trigger-characters, however I am a little scared to make a change late in the endgame.

**Comment by jrieken** [comment]
Ok, the competing provider is the html-extension that registers those completion item provider: https://github.com/Microsoft/vscode/blob/master/extensions/html/server/src/htmlServerMain.ts#L75. Still not sure what as changed but I'll think about a fix

**Comment by jrieken** [comment]
@ramya-rao-a I do have a change ready that makes sure to *also* resolve incomplete suggest results when a trigger character is being hit. However, when I ask emmet for completions with the `html:` prefix I just get that back, so no `html:5`. Fixing this bug uncovers another bug... 

I believe that this used to work by accident (more than one thing is off here) and that's because before the completion item would have a filter `html:5` and now it's just `html`. Both get compared to `html:` and latter won't match because it's shorter. 

<img width="752" alt="screen shot 2017-06-29 at 14 26 24" src="https://user-images.githubusercontent.com/1794099/27687621-ccf64034-5cd7-11e7-8e27-d3a62d7a779c.png">

<img width="586" alt="screen shot 2017-06-29 at 14 28 27" src="https://user-images.githubusercontent.com/1794099/27687622-cd1b9b04-5cd7-11e7-9cc2-8aad4068c628.png">

**Comment by ramya-rao-a** [comment]
Thanks for all the digging @jrieken 
I agree its too late in the end game for risky fixes.

> I believe that this used to work by accident

I have added `:` as trigger character for emmet completions and updated the completion provider to return expected items. 
So after https://github.com/Microsoft/vscode/commit/42ae1fc0f6122e5ca6b5237622bc8adadbb1c4f2 and https://github.com/ramya-rao-a/vscode-emmet-helper/commit/5875048b62a907eb4e7ebcbab7811ce878cb3f68, we get expected outcome.

**Comment by ramya-rao-a** [comment]
Still an issue for abbreviations like `ul*2`
The suggestions disappear when * is typed.

Re-opening

**Comment by jrieken** [comment]
Yeah, whenever we hit a word stops with incomplete results and when the result isn't matching that we cancel. That was done in https://github.com/Microsoft/vscode/commit/5ad16183ab248948fdc457af0340321caf1ba987. Feel free to revert and reopen the corresponding issue. I am on vacation now... 🏖

**Comment by jens1o** [comment]
Happy holidays @jrieken 😃

**Comment by ramya-rao-a** [comment]
I add the * and $ to trigger char list as well. So we are good on those fronts.

**Comment by ramya-rao-a** [comment]
Note to verifier:
- Try abbreivations with * $ : in your abbreviations and check you dont lose the suggestion list
- Refer to https://docs.emmet.io/cheat-sheet/
- Few things from the cheat sheet is not working due to upstream issue. See https://github.com/emmetio/atom-plugin/issues/27

**Comment by nanaasumadu** [comment]
@ramya-rao-a 

Just tested emmett suggestions with `:` and `*` and there work fine.  `$` also works in my php file but I am not sure if that's coming from emmet or the language server.

Thank you so much for your great work.

**Comment by ramya-rao-a** [comment]
Thanks for all the feedback @nanaasumadu and @jens1o
